### PR TITLE
[7 of #107] expand env parsing and command creation 

### PR DESCRIPTION
**Pull request by bollwyvl**
This collapses the boolean logic for parsing an environment file, and adds the parsed environment (as yaml or explicit lines) to the dynamic options. Subsequently, it moves the "create an env" logic to dedicated files for "simple" conda create, create with a lockfile and `env update` with a YAML. This drops a _lot_ from `setup.ts`, and notably a number of transient dependencies (`yaml`, etc).

- continues #107
- includes
  - #126  
- might help #122?

## On the next PR

> refactor the tools provider

Apply a similar treatment as this PR to the provisioning of tools in the base env (e.g. `mamba`, `conda-build`), prior to installation of the environment.

Again, we need a way to say, _hey, my base env already/only has `mamba`_. One Does Not Simply Parse a Conda MatchSpec, though... so I guess I'd rather it was explicit, e.g. `installer-has-mamba`, which we'd be able to pre-populate for `mambaForge` rather than trying to do anything with `mamba-version`.

### CODE REVIEW

**Comment by bollwyvl** [comment]
Diff's getting a little nasty, will keep this `wip` until #126 lands...

**Comment by goanpeca** [comment]
#126 landed @bollwyvl

File: src/types.ts [hunk 1]
```diff
@@ -125,3 +125,24 @@ export interface IInstallerProvider {
     options: IDynamicOptions
   ) => Promise<IInstallerResult>;
 }
+
+/**
+ * A strategy for ensure an environment is up-to-date vs the action inputs
```
Comment (bollwyvl) [inline src/types.ts:130]: typo, will batch with other review fixes

File: src/conda.ts [hunk 2]
```diff
@@ -62,47 +62,63 @@ export async function bootstrapConfig(): Promise<void> {
...
-          );
-        } catch (err) {
-          core.warning(`Couldn't set conda configuration '${key}'`);
-        }
-      }
-    }
+  // channels are special: if specified as an action input, these take priority
+  let channels = inputs.condaConfig.channels
+    .trim()
+    .split(/[,\n]/)
```
Comment (bollwyvl) [inline src/conda.ts:92]: this inadvertently allows for `\n` in addition to `,` as a channel separator. having both is probably a bug unless we document it... will revert to existing behavior.

the correct behavior is probably to _only_ use `\n` _if_ there is a (non-trailing?) `\n` in the value

Comment (goanpeca) [reply src/conda.ts:92]: Sounds good. The correct behavior to me would be to have all commas or all line separators.

Could we do something like `/(,)|(\n)/`and count each group?

File: src/conda.ts [hunk 3]
```diff
@@ -62,47 +62,63 @@ export async function bootstrapConfig(): Promise<void> {
   );
 }
 
+/**
+ * Copy the given condarc file into place
+ */
+export async function copyConfig(inputs: types.IActionInputs) {
+  const sourcePath: string = path.join(
+    process.env["GITHUB_WORKSPACE"] || "",
```
Comment (goanpeca) [inline src/conda.ts:70]: Should we use the home path?

File: src/conda.ts [hunk 4]
```diff
@@ -62,47 +62,63 @@ export async function bootstrapConfig(): Promise<void> {
...
+  }
+
+  // LIFO: reverse order to preserve higher priority as listed in the option
+  for (const channel of channels.reverse()) {
+    core.info(`Adding channel '${channel}'`);
+    await condaCommand(["config", "--add", "channels", channel], options);
   }
 
-  await condaCommand(["config", "--show-sources"], notMambaOptions);
+  // all other options
```
Comment (goanpeca) [inline src/conda.ts:106]: Feel like all these comments should be capitalized, unless there is some sort of TS convention.

File: src/constants.ts [hunk 5]
```diff
@@ -66,3 +66,21 @@ export const CONDA_CACHE_FOLDER = "conda_pkgs_dir";
...
+ *
+ * ### Note
+ * Some examples:
+ * - python
+ * - python 3
+ * - python>3
+ * - python!=2
+ * - conda-forge::python
+ *
+ * TODO: this should be generalized, and, along with roundtrip parsing/generating
```
Comment (goanpeca) [inline src/constants.ts:82]: Yeah, could be :)

File: src/utils.ts [hunk 6]
```diff
@@ -48,3 +48,17 @@ export async function execute(command: string[]): Promise<void> {
     throw new Error(`${command[0]} return error code ${rc}`);
   }
 }
+
+/**
+ * create a conda veersion spec string.
```
Comment (goanpeca) [inline src/utils.ts:53]: Typo

**Review by goanpeca — CHANGES_REQUESTED** [review]
Couple of questions, and also I feel like all comments should be capitalized 🤷🏽 

Thanks for this amazing work @bollwyvl !

File: src/conda.ts (hunk 3 above)
Comment (bollwyvl) [reply src/conda.ts:70]: I think we need to assume sourcing the config from the workspace, which may or may not be in home.

the `CONDARC_PATH` on the other hand _is_ in `HOME`, as reported by `os.homedir()`

File: src/conda.ts (hunk 2 above)
Comment (bollwyvl) [reply src/conda.ts:92]: As i said, let's not monkey with the newlines on this PR, will put it back to comma-only.

If, for some awful reason, someone _must_ have a `,` in their URL, as it _is_ part of the RFC, allowing both is going to break, hard. 

Typographically, just commas scans quickly horizontally, while the lines look more like all the other conda docs. I don't see a huge win to supporting both in one, either:
```yaml
channels: |
  conda-forge,defaults
  https://www.crazy.town/conda-remote/
```
one must really parse what each line is saying, and there's no whitespace to help...

Comment (goanpeca) [reply src/conda.ts]: Ok, commas for now then :)

**Comment by bollwyvl** [comment]
Hopefully all typos/capitalization handled on
https://github.com/conda-incubator/setup-miniconda/pull/127/commits/64e43746f1144df89ec0c932169bfaec527fe6b3

**Review by goanpeca — APPROVED** [review]

**Pull request merged**